src/community/wfs3/src/main/java/org/geoserver/wfs3/response/RFCGeoJSONFeaturesResponse.java

```diff
@@ -0,0 +1,189 @@
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.wfs3.response;
+
+import org.geoserver.catalog.FeatureTypeInfo;
+import org.geoserver.config.GeoServer;
+import org.geoserver.ows.Dispatcher;
+import org.geoserver.ows.Request;
+import org.geoserver.ows.URLMangler;
+import org.geoserver.ows.util.ResponseUtils;
+import org.geoserver.platform.Operation;
+import org.geoserver.platform.ServiceException;
+import org.geoserver.wfs.json.GeoJSONBuilder;
+import org.geoserver.wfs.json.GeoJSONGetFeatureResponse;
+import org.geoserver.wfs.request.FeatureCollectionResponse;
+import org.geoserver.wfs.request.GetFeatureRequest;
+import org.geoserver.wfs.request.Query;
+import org.geoserver.wfs3.NCNameResourceCodec;
+import org.geotools.feature.FeatureCollection;
+import org.geotools.feature.NameImpl;
+import org.geotools.referencing.CRS;
+import org.geotools.referencing.crs.DefaultGeographicCRS;
+import org.geotools.util.Version;
+import org.opengis.feature.Feature;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+
+import javax.xml.namespace.QName;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.UnsupportedEncodingException;
+import java.math.BigInteger;
+import java.util.Collections;
+import java.util.List;
+
+import static org.geoserver.wfs3.DefaultWebFeatureService30.getAvailableFormats;
+
+public class RFCGeoJSONFeaturesResponse extends GeoJSONGetFeatureResponse {
+
+    static final ThreadLocal<String> WFS3_FEATURE_ID = new ThreadLocal<String>();
+
+    /** The MIME type requested by WFS3 for GeoJSON Responses */
+    public static final String MIME = "application/geo+json";
+
+    public RFCGeoJSONFeaturesResponse(GeoServer gs) {
+        super(gs, MIME);
+    }
+
+    @Override
+    public String getMimeType(Object value, Operation operation) throws ServiceException {
+        return MIME;
+    }
+
+    public void write(Object value, OutputStream output, Operation operation) throws IOException {
+        // was it a single feature request?
+        String requestFeatureId = getWFS3FeatureId();
+        if (requestFeatureId != null) {
+            try {
+                WFS3_FEATURE_ID.set(requestFeatureId);
+                writeSingleFeature((FeatureCollectionResponse) value, output, operation);
+            } finally {
+                WFS3_FEATURE_ID.remove();
+            }
+        } else {
+            super.write(value, output, operation);
+        }
+    }
+
+    /**
+     * Returns the WFS3 featureId, or null if it's missing or the request is not a WFS3 one
+     *
+     * @return
+     */
+    private String getWFS3FeatureId() {
+        Request dr = Dispatcher.REQUEST.get();
+        String featureId = null;
+        if (dr != null && (new Version(dr.getVersion()).getMajor().equals(3))) {
+            Object featureIdValue = dr.getKvp().get("featureId");
```

Comment:
The case sensitive looks weird here, the WFS3 query parameters are case sensitive ?